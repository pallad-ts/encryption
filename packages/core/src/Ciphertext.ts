import { Either, fromTry, left, right } from "@sweet-monads/either";
import { KeyId, KeyIdSchema } from "@pallad/keyring";
import { ERRORS } from "./errors";
import { z } from "zod";
import { Buffer } from "node:buffer";
import { EncryptionError } from "./EncryptionError";

const MAX_ENCRYPTED_SIZE = 2 ** 16 - 1;
const schema = z
    .object({
        keyId: KeyIdSchema,
        iv: z
            .instanceof(Buffer)
            .refine(value => value.length > 0, "IV cannot be empty")
            .refine(value => value.length < 256, "IV is too long"),
        encrypted: z
            .instanceof(Buffer)
            .refine(value => value.length > 0, "Encrypted data cannot be empty")
            .refine(value => value.length < MAX_ENCRYPTED_SIZE, "Encrypted data is too long"),
    })
    .transform(({ keyId, iv, encrypted }) => {
        return Object.freeze(
            Object.create(Ciphertext.prototype, {
                keyId: { value: keyId, enumerable: true },
                iv: { value: iv, enumerable: true },
                encrypted: { value: encrypted, enumerable: true },
            })
        ) as Ciphertext;
    });

export class Ciphertext {
    readonly keyId: KeyId;
    readonly iv: Buffer;
    readonly encrypted: Buffer;

    constructor() {
        throw new Error("Use Ciphertext.fromString or Ciphertext.schema.parse to create Ciphertext");
    }

    toString() {
        return Buffer.concat(Array.from(encode(this))).toString("base64");
    }

    static fromString(input: string): Either<string, Ciphertext> {
        return fromTry<EncryptionError, Ciphertext>(() => {
            const [keyId, iv, encrypted] = Array.from(decode(input));
            return Ciphertext.schema.parse({ keyId, iv, encrypted });
        }).mapLeft(error => error.message);
    }

    static schema = schema;
}

function* encode(input: Ciphertext): Generator<Buffer> {
    yield Buffer.from([input.keyId.length]);
    yield Buffer.from(input.keyId, "utf-8");
    yield Buffer.from([input.iv.length]);
    yield input.iv;
    const encryptedSizeBuffer = Buffer.alloc(2);
    encryptedSizeBuffer.writeUInt16BE(input.encrypted.length);
    yield encryptedSizeBuffer;
    yield input.encrypted;
}

const MALFORMED = "Malformed ciphertext";

function* decode(input: string) {
    const buffer = Buffer.from(input, "base64");

    let offset = 0;
    const keySize = readUInt8(buffer, offset++);
    yield readAscii(buffer, offset, keySize);
    offset += keySize;

    const ivSize = readUInt8(buffer, offset++);
    yield readBuffer(buffer, offset, ivSize);
    offset += ivSize;

    const encryptedSize = readUInt16(buffer, offset);
    offset += 2;
    yield readBuffer(buffer, offset, encryptedSize);
    offset += encryptedSize;

    if (offset !== buffer.length) {
        throw new EncryptionError("Invalid input length");
    }
}

function readUInt8(input: Buffer, offset: number) {
    if (input.length < offset + 1) {
        throw new EncryptionError(MALFORMED);
    }
    return input.readUInt8(offset);
}

function readUInt16(input: Buffer, offset: number) {
    if (input.length < offset + 2) {
        throw new EncryptionError(MALFORMED);
    }
    return input.readUInt16BE(offset);
}

function readAscii(input: Buffer, offset: number, length: number) {
    if (input.length < offset + length) {
        throw new EncryptionError(MALFORMED);
    }
    return input.toString("ascii", offset, offset + length);
}

function readBuffer(input: Buffer, offset: number, length: number) {
    if (input.length < offset + length) {
        throw new EncryptionError(MALFORMED);
    }
    return input.subarray(offset, offset + length);
}
