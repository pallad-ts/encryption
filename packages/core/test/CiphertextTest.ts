import { Ciphertext } from "@src/Ciphertext";
import { randomBytes } from "crypto";
import "@pallad/errors-dev";

describe("Ciphertext", () => {
    it("encoding", () => {
        const ciphertext = Ciphertext.schema.parse({
            keyId: "keyId",
            iv: Buffer.from("some_iv"),
            encrypted: Buffer.from("some_encrypted"),
        });

        expect(ciphertext.toString()).toEqual("BWtleUlkB3NvbWVfaXYADnNvbWVfZW5jcnlwdGVk");
    });

    describe("creating", () => {
        it("cannot be instantiated directly", () => {
            expect(() => {
                new Ciphertext();
            }).toThrowErrorMatchingSnapshot();
        });

        describe("from string", () => {
            describe("fail", () => {
                function* cases(): Generator<[string, Buffer]> {
                    const KEY = "some_key_id";
                    const bufferList: Buffer[] = [];

                    // key
                    yield ["empty string", Buffer.concat(bufferList)];
                    bufferList.push(Buffer.from([KEY.length]));
                    yield ["missing key id", Buffer.concat(bufferList)];
                    yield ["invalid key id length", Buffer.concat([...bufferList, randomBytes(4)])];
                    bufferList.push(Buffer.from(KEY, "ascii"));

                    // iv
                    yield ["missing iv length", Buffer.concat(bufferList)];
                    bufferList.push(Buffer.from([12]));
                    yield ["missing iv", Buffer.concat(bufferList)];
                    yield ["invalid iv length", Buffer.concat([...bufferList, randomBytes(4)])];
                    bufferList.push(randomBytes(12));

                    // encrypted
                    yield ["missing encrypted length", Buffer.concat(bufferList)];
                    const encryptedLength = Buffer.alloc(2);
                    encryptedLength.writeUInt16BE(50);
                    bufferList.push(encryptedLength);
                    yield ["missing encrypted", Buffer.concat(bufferList)];
                    yield ["invalid encrypted length", Buffer.concat([...bufferList, randomBytes(4)])];
                    bufferList.push(randomBytes(50));

                    // extra data
                    yield ["extra data", Buffer.concat([...bufferList, randomBytes(10)])];
                }

                it.each(Array.from(cases()))("malformed %s", (name, buffer) => {
                    const result = Ciphertext.fromString(buffer.toString("base64"));
                    expect(result.isLeft()).toBe(true);
                    expect(result.value).toMatchSnapshot();
                });
            });

            it("success", () => {
                const iv = randomBytes(10);
                const encrypted = randomBytes(50);
                const ciphertext = Ciphertext.schema.parse({ keyId: "key", iv, encrypted });

                const result = Ciphertext.fromString(ciphertext.toString());
                expect(result.value).toMatchObject({
                    keyId: "key",
                    iv: iv,
                    encrypted: encrypted,
                });
            });
        });
    });
});
