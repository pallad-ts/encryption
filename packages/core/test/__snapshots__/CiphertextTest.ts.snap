// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ciphertext creating cannot be instantiated directly 1`] = `"Use Ciphertext.fromString or Ciphertext.schema.parse to create Ciphertext"`;

exports[`Ciphertext creating from string fail malformed empty string 1`] = `"Malformed ciphertext"`;

exports[`Ciphertext creating from string fail malformed extra data 1`] = `"Invalid input length"`;

exports[`Ciphertext creating from string fail malformed invalid encrypted length 1`] = `"Malformed ciphertext"`;

exports[`Ciphertext creating from string fail malformed invalid iv length 1`] = `"Malformed ciphertext"`;

exports[`Ciphertext creating from string fail malformed invalid key id length 1`] = `"Malformed ciphertext"`;

exports[`Ciphertext creating from string fail malformed missing encrypted 1`] = `"Malformed ciphertext"`;

exports[`Ciphertext creating from string fail malformed missing encrypted length 1`] = `"Malformed ciphertext"`;

exports[`Ciphertext creating from string fail malformed missing iv 1`] = `"Malformed ciphertext"`;

exports[`Ciphertext creating from string fail malformed missing iv length 1`] = `"Malformed ciphertext"`;

exports[`Ciphertext creating from string fail malformed missing key id 1`] = `"Malformed ciphertext"`;
